File: tablet_qt/tasks/apeqpt.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

// By Joe Kearney, Rudolf Cardinal.

#include "apeqpt.h"

#include "lib/datetime.h"
#include "lib/stringfunc.h"
#include "lib/uifunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/qudatetime.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/questionnairefunc.h"
#include "questionnairelib/qumcq.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "questionnairelib/qutextedit.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"

using mathfunc::anyNullOrEmpty;
using mathfunc::sumInt;
using mathfunc::totalScorePhrase;
using stringfunc::strseq;

const QString Apeqpt::APEQPT_TABLENAME("apeqpt");

const QString FN_DATETIME("q_datetime");

const QString CHOICE_SUFFIX("_choice");
const QString SAT_SUFFIX("_satisfaction");

const int CHOICE_QUESTIONS_N = 3;

const QString FN_Q1_SATISFACTION("q1_satisfaction");
const QString FN_Q2_SATISFACTION("q2_satisfaction");

void initializeApeqpt(TaskFactory& factory)
{
    static TaskRegistrar<Apeqpt> registered(factory);
}

Apeqpt::Apeqpt(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, APEQPT_TABLENAME, true, false, false),
    // ... anon, clin, resp
    m_questionnaire(nullptr)
{
    addField(FN_DATETIME, QMetaType::fromType<QDateTime>());

    for (const QString& field :
         strseq("q", 1, CHOICE_QUESTIONS_N, CHOICE_SUFFIX)) {
        addField(field, QMetaType::fromType<int>());
    }

    addField(FN_Q1_SATISFACTION, QMetaType::fromType<int>());
    addField(FN_Q2_SATISFACTION, QMetaType::fromType<QString>());

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.

    // Extra initialization:
    if (load_pk == dbconst::NONEXISTENT_PK) {
        setValue(FN_DATETIME, datetime::now(), false);
    }
}

// ============================================================================
// Class info
// ============================================================================

QString Apeqpt::shortname() const
{
    return "APEQPT";
}

QString Apeqpt::longname() const
{
    return tr(
        "Assessment Patient Experience Questionnaire "
        "for Psychological Therapies"
    );
}

QString Apeqpt::description() const
{
    return tr(
        "Patient feedback questionnaire on assessment for psychological "
        "therapy/choosing treatment."
    );
}

// ============================================================================
// Instance info
// ============================================================================

bool Apeqpt::isComplete() const
{
    QStringList required_always{FN_Q1_SATISFACTION};
    for (const QString& field :
         strseq("q", 1, CHOICE_QUESTIONS_N, CHOICE_SUFFIX)) {
        required_always.append(field);
    }
    required_always.append(FN_DATETIME);
    return !anyNullOrEmpty(values(required_always));
}

NameValueOptions Apeqpt::optionsSatisfaction() const
{
    return NameValueOptions{
        {xstring("a4_satisfaction"), 4},  // Completely satisfied
        {xstring("a3_satisfaction"), 3},
        {xstring("a2_satisfaction"), 2},
        {xstring("a1_satisfaction"), 1},
        {xstring("a0_satisfaction"), 0},  // Not at all satisfied
    };
}

NameValueOptions Apeqpt::optionsChoiceWithNA() const
{
    return NameValueOptions{
        {xstring("a1_choice"), 1},  // Yes
        {xstring("a0_choice"), 0},  // No
        {xstring("a2_choice"), 2},  // N/A
    };
}

QStringList Apeqpt::summary() const
{
    const NameValueOptions options_satisfaction = optionsSatisfaction();
    return QStringList{
        QString("Patient Satisfaction: %1")
            .arg(options_satisfaction.nameFromValue(value(FN_Q1_SATISFACTION))
            )};
}

QStringList Apeqpt::detail() const
{
    QStringList lines = completenessInfo();
    const NameValueOptions ans = optionsChoiceWithNA();
    const QString spacer = " ";
    QString summaryLine, xstringname, fieldname, qnum;
    lines.append("<b>Choice</b>:");
    for (int i = 0; i < CHOICE_QUESTIONS_N; ++i) {
        qnum = QString::number(i + 1);
        xstringname = QString("q%1_choice_s").arg(qnum);
        fieldname = QString("q%1_choice").arg(qnum);
        summaryLine = QString("Q%1 %2: %3")
                          .arg(
                              qnum,
                              xstring(xstringname),
                              ans.nameFromValue(value(fieldname))
                          );
        lines.append(summaryLine);
    }
    lines.append("");
    lines.append("<b>Satisfaction</b>:");
    lines.append(summary());
    lines.append("");
    lines.append("<b>Additional feedback</b>:");
    lines.append(value(FN_Q2_SATISFACTION).toString());
    return lines;
}

OpenableWidget* Apeqpt::editor(const bool read_only)
{
    const NameValueOptions options_choice{
        {xstring("a1_choice"), 1},  // Yes
        {xstring("a0_choice"), 0},  // No
    };
    const NameValueOptions options_choice_with_na = optionsChoiceWithNA();
    const NameValueOptions options_satisfaction = optionsSatisfaction();

    const int question_width = 25;
    const QVector<int> yes_no_opts_widths{38, 37};
    const QVector<int> all_opts_widths{25, 25, 25};

    QuPagePtr page(new QuPage{
        (new QuText(xstring("instructions_to_subject_1")))
            ->setItalic()
            ->setBig(),
        (new QuText(xstring("instructions_to_subject_2")))
            ->setItalic()
            ->setBig(),
        (new QuText(xstring("q_date")))->setBold(),
        (new QuDateTime(fieldRef(FN_DATETIME)))->setOfferNowButton(true),
        (new QuText(xstring("h1")))->setBig()->setBold(),
        (new QuMcqGrid(
             {
                 QuestionWithOneField(
                     xstring("q1_choice"), fieldRef("q1_choice")
                 ),
                 QuestionWithOneField(
                     xstring("q2_choice"), fieldRef("q2_choice")
                 ),
             },
             options_choice
         ))
            ->setWidth(question_width, yes_no_opts_widths)
            ->setExpand(true),
        (new QuMcqGrid(
             {
                 QuestionWithOneField(
                     xstring("q3_choice"), fieldRef("q3_choice")
                 ),
             },
             options_choice_with_na
         ))
            ->setWidth(question_width, all_opts_widths)
            ->setExpand(true),
        (new QuText(xstring("h2")))->setBig()->setBold(),
        (new QuMcq(fieldRef(FN_Q1_SATISFACTION), options_satisfaction))
            ->setHorizontal(true)
            ->setAsTextButton(true),
        (new QuText(xstring("q2_satisfaction")))->setBold(),
        new QuTextEdit(fieldRef(FN_Q2_SATISFACTION, false)),
        (new QuText(xstring("thanks")))->setItalic(),
    });

    page->setTitle(longname());

    m_questionnaire = new Questionnaire(m_app, {page});

    m_questionnaire->setReadOnly(read_only);

    return m_questionnaire;
}

File: server/camcops_server/tasks/apeqpt.py
#!/usr/bin/env python

"""
camcops_server/tasks/apeqpt.py

===============================================================================

    Copyright (C) 2012-2020 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

- By Joe Kearney, Rudolf Cardinal.

"""

from typing import Dict, List, TYPE_CHECKING

from fhirclient.models.questionnaire import (
    QuestionnaireItem,
    QuestionnaireItemAnswerOption,
)
from fhirclient.models.questionnaireresponse import (
    QuestionnaireResponseItem,
    QuestionnaireResponseItemAnswer,
)
from sqlalchemy.sql.sqltypes import Integer, UnicodeText

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_html import tr_qa
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    CamcopsColumn,
    PendulumDateTimeAsIsoTextColType,
    ZERO_TO_ONE_CHECKER,
    ZERO_TO_TWO_CHECKER,
    ZERO_TO_FOUR_CHECKER
)
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import (
    get_from_dict,
    Task,
)

if TYPE_CHECKING:
    from camcops_server.cc_modules.cc_exportrecipient import ExportRecipient


# =============================================================================
# APEQPT
# =============================================================================

class Apeqpt(Task):
    """
    Server implementation of the APEQPT task.
    """
    __tablename__ = "apeqpt"
    shortname = "APEQPT"
    provides_trackers = True

    q_datetime = CamcopsColumn(
        "q_datetime", PendulumDateTimeAsIsoTextColType,
        comment="Date/time the assessment tool was completed")

    N_CHOICE_QUESTIONS = 3
    q1_choice = CamcopsColumn(
        "q1_choice", Integer,
        comment="Enough information was provided (0 no, 1 yes)",
        permitted_value_checker=ZERO_TO_ONE_CHECKER)
    q2_choice = CamcopsColumn(
        "q2_choice", Integer,
        comment="Treatment preference (0 no, 1 yes)",
        permitted_value_checker=ZERO_TO_ONE_CHECKER)
    q3_choice = CamcopsColumn(
        "q3_choice", Integer,
        comment="Preference offered (0 no, 1 yes, 2 N/A)",
        permitted_value_checker=ZERO_TO_TWO_CHECKER)

    q1_satisfaction = CamcopsColumn(
        "q1_satisfaction", Integer,
        comment=(
            "Patient satisfaction (0 not at all satisfied - "
            "4 completely satisfied)"
        ),
        permitted_value_checker=ZERO_TO_FOUR_CHECKER)
    q2_satisfaction = CamcopsColumn(
        "q2_satisfaction", UnicodeText,
        comment="Service experience")

    MAIN_QUESTIONS = [
        "q_datetime",
        "q1_choice",
        "q2_choice",
        "q3_choice",
        "q1_satisfaction",
    ]

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Assessment Patient Experience Questionnaire "
                 "for Psychological Therapies")

    def is_complete(self) -> bool:
        if self.any_fields_none(self.MAIN_QUESTIONS):
            return False
        if not self.field_contents_valid():
            return False
        return True

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields()

    def get_task_html(self, req: CamcopsRequest) -> str:
        c_dict = {
            0: "0 — " + self.wxstring(req, "a0_choice"),
            1: "1 — " + self.wxstring(req, "a1_choice"),
            2: "2 — " + self.wxstring(req, "a2_choice"),
        }
        s_dict = {
            0: "0 — " + self.wxstring(req, "a0_satisfaction"),
            1: "1 — " + self.wxstring(req, "a1_satisfaction"),
            2: "2 — " + self.wxstring(req, "a2_satisfaction"),
            3: "3 — " + self.wxstring(req, "a3_satisfaction"),
            4: "4 — " + self.wxstring(req, "a4_satisfaction"),
        }
        q_a = ""
        for i in range(1, self.N_CHOICE_QUESTIONS + 1):
            nstr = str(i)
            q_a += tr_qa(
                self.wxstring(req, "q" + nstr + "_choice"),
                get_from_dict(c_dict, getattr(self, "q" + nstr + "_choice")))

        q_a += tr_qa(self.wxstring(req, "q1_satisfaction"),
                     get_from_dict(s_dict, self.q1_satisfaction))
        q_a += tr_qa(self.wxstring(req, "q2_satisfaction"),
                     self.q2_satisfaction, default="")

        return f"""
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {self.get_is_complete_tr(req)}
                </table>
            </div>
            <div class="{CssClass.EXPLANATION}">
                Patient satisfaction rating for service provided. The service
                is rated on choice offered and general satisfaction.
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="60%">Question</th>
                    <th width="40%">Answer</th>
                </tr>
                {q_a}
            </table>
        """

    def get_fhir_questionnaire_items(
            self,
            req: "CamcopsRequest",
            recipient: "ExportRecipient") -> List[Dict]:
        items = [
            QuestionnaireItem(jsondict={
                "linkId": "q_datetime",
                "text": self.wxstring(req, "q_date"),
                "type": "dateTime",
            }).as_json()
        ]

        yes_no_options = []

        for index in range(2):
            yes_no_options.append(QuestionnaireItemAnswerOption(jsondict={
                "valueCoding": {
                    "code": str(index),
                    "display": self.wxstring(req, f"a{index}_choice"),
                }
            }).as_json())

        items.append(QuestionnaireItem(jsondict={
            "linkId": "q1_choice",
            "text": self.wxstring(req, "q1_choice"),
            "type": "choice",
            "answerOption": yes_no_options,
        }).as_json())

        items.append(QuestionnaireItem(jsondict={
            "linkId": "q2_choice",
            "text": self.wxstring(req, "q2_choice"),
            "type": "choice",
            "answerOption": yes_no_options,
        }).as_json())

        yes_no_na_options = yes_no_options + [
            QuestionnaireItemAnswerOption(
                jsondict={
                    "valueCoding": {
                        "code": "2",
                        "display": self.wxstring(req, "a2_choice"),
                    }
                }
            ).as_json()
        ]

        items.append(QuestionnaireItem(jsondict={
            "linkId": "q3_choice",
            "text": self.wxstring(req, "q3_choice"),
            "type": "choice",
            "answerOption": yes_no_na_options,
        }).as_json())

        satisfaction_options = []

        for index in range(5):
            satisfaction_options.append(QuestionnaireItemAnswerOption(jsondict={
                "valueCoding": {
                    "code": str(index),
                    "display": self.wxstring(req, f"a{index}_satisfaction"),
                }
            }).as_json())

        items.append(QuestionnaireItem(jsondict={
            "linkId": "q1_satisfaction",
            "text": self.wxstring(req, "q1_satisfaction"),
            "type": "choice",
            "answerOption": satisfaction_options,
        }).as_json())

        items.append(QuestionnaireItem(jsondict={
            "linkId": "q2_satisfaction",
            "text": self.wxstring(req, "q2_satisfaction"),
            "type": "string",
        }).as_json())

        return items

    def get_fhir_questionnaire_response_items(
            self,
            req: "CamcopsRequest",
            recipient: "ExportRecipient") -> List[Dict]:

        items = []

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueDateTime": self.q_datetime.isoformat()
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q_datetime",
            "text": self.wxstring(req, "q_date"),
            "answer": [answer.as_json()],
        }).as_json())

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueInteger": self.q1_choice
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q1_choice",
            "text": self.wxstring(req, "q1_choice"),
            "answer": [answer.as_json()],
        }).as_json())

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueInteger": self.q2_choice
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q2_choice",
            "text": self.wxstring(req, "q2_choice"),
            "answer": [answer.as_json()],
        }).as_json())

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueInteger": self.q3_choice
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q3_choice",
            "text": self.wxstring(req, "q3_choice"),
            "answer": [answer.as_json()],
        }).as_json())

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueInteger": self.q1_satisfaction
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q1_satisfaction",
            "text": self.wxstring(req, "q1_satisfaction"),
            "answer": [answer.as_json()],
        }).as_json())

        answer = QuestionnaireResponseItemAnswer(jsondict={
            "valueString": self.q2_satisfaction
        })
        items.append(QuestionnaireResponseItem(jsondict={
            "linkId": "q2_satisfaction",
            "text": self.wxstring(req, "q2_satisfaction"),
            "answer": [answer.as_json()],
        }).as_json())

        return items
